File: src/com/implimentz/fragments/AnnotationManager.java
package info.wikiroutes.android.fragments;

import android.util.NoSuchPropertyException;


/**
 * Created by Alexander Efremenkov.
 * Date: 06.01.2016, 16:01.
 * In Intellij IDEA 14.1 Ultimate
 * aefremenkov@livamster.ru
 */
final class AnnotationManager {

    private static final String META_EXCEPTION = "Fragment %1$s must contain %2$s annotation";

    final FragmentMeta getFragmentMeta(final Fragment fragment) {
        final FragmentMeta annotation = fragment.getClass().getAnnotation(FragmentMeta.class);
        if(annotation == null) {
            throw new NoSuchPropertyException(
                    String.format(META_EXCEPTION, fragment.getClass().getSimpleName(), FragmentMeta.class.getSimpleName())
            );
        }
        return annotation;
    }
}

File: src/main/java/com/implimentz/fragments/AnnotationManager.kt
package com.implimentz.fragments

import android.support.v7.app.AppCompatActivity
import android.util.NoSuchPropertyException


/**
 * Created by ironz.
 * Date: 06.01.2016, 16:01.
 * In Intellij IDEA 15.0.3 Ultimate
 * email: implimentz@gmail.com
 * twitter: iamironz
 */
internal object AnnotationManager {

    fun <D, A : AppCompatActivity> getMeta(fragment: Fragment<D, A>): FragmentMeta {

        val clazz = FragmentMeta::class.java
        val simpleName: String = fragment.javaClass.simpleName

        val annotation: FragmentMeta? = fragment.javaClass.getAnnotation(clazz)

        return annotation ?: throw NoSuchPropertyException(
                "Fragment $simpleName must contain ${clazz.name} annotation"
        )
    }

    fun <D, A : AppCompatActivity> getLayoutId(fragment: Fragment<D, A>): Int {

        val clazz: Class<FragmentLayout> = FragmentLayout::class.java
        val simpleName: String = fragment.javaClass.simpleName

        val annotation: FragmentLayout? = fragment.javaClass.getAnnotation(clazz)

        if (annotation != null) {
            return annotation.value
        }

        throw NoSuchPropertyException("Fragment $simpleName must contain ${clazz.name} annotation")
    }

    fun <D, A : AppCompatActivity> getMenuId(fragment: Fragment<D, A>): Int {

        val clazz: Class<FragmentMenu> = FragmentMenu::class.java

        val annotation: FragmentMenu? = fragment.javaClass.getAnnotation(clazz)

        return if (annotation == null) 0 else annotation.value
    }
}
